lib/ansible/cli/scripts/ansible_connection_cli_stub.py

```diff
@@ -132,31 +132,36 @@ def run(self):
 
                 self.exception = None
                 (s, addr) = self.sock.accept()
-                signal.alarm(0)
-                signal.signal(signal.SIGALRM, self.command_timeout)
-                while True:
-                    data = recv_data(s)
-                    if not data:
-                        break
+                try:
+                    signal.alarm(0)
+                    signal.signal(signal.SIGALRM, self.command_timeout)
+                    while True:
+                        data = recv_data(s)
+                        if not data:
+                            break
 
-                    if log_messages:
-                        display.display("jsonrpc request: %s" % data, log_only=True)
+                        if log_messages:
+                            display.display("jsonrpc request: %s" % data, log_only=True)
 
-                    request = json.loads(to_text(data, errors='surrogate_or_strict'))
-                    if request.get('method') == "exec_command" and not self.connection.connected:
-                        self.connection._connect()
+                        request = json.loads(to_text(data, errors='surrogate_or_strict'))
+                        if request.get('method') == "exec_command" and not self.connection.connected:
+                            self.connection._connect()
 
-                    signal.alarm(self.connection.get_option('persistent_command_timeout'))
+                        signal.alarm(self.connection.get_option('persistent_command_timeout'))
 
-                    resp = self.srv.handle_request(data)
-                    signal.alarm(0)
+                        resp = self.srv.handle_request(data)
+                        signal.alarm(0)
 
-                    if log_messages:
-                        display.display("jsonrpc response: %s" % resp, log_only=True)
+                        if log_messages:
+                            display.display("jsonrpc response: %s" % resp, log_only=True)
 
-                    send_data(s, to_bytes(resp))
+                        send_data(s, to_bytes(resp))
 
-                s.close()
+                except Exception as exc:
```

Comment:
The purpose of catching Exception here is that it is the catch of last resort for anything that goes wrong in the plugins. This change forwards that error to the controller, whereas without this `ansible-connection` just dies and the module calling the persistent connection never finds out why.

This is not a common scenario, but if someone is doing development on a plugin in the persistent connection space and fails to account for an exception on the remote process, there is presently no good way to figure out what is wrong, and this is intended to fix that.

Reply:
Reraised to... where, exactly? This code runs in a fork of a subprocess. Forwarding the exception over the socket is the only way we _can_ let the outer layers react, because what happens immediately after is that all the connections are closed and the fork is terminated. The outer `try` block is just for handling things that go wrong before we have a socket opened back to the main ansible process... socket errors or timeouts, and SIGTERM. The inner `try` is for anything that happens once we're sure the socket back can be used to say what went wrong. And we should do this because the other side of that socket is expecting a response to its request and not just the connection to be closed with no information as happens now.

The later `shutdown()` still happens regardless of where an exception is raised. That's why it, along with `s.close()` here are in `finally` blocks, to make sure that the cleanup happens properly regardless of whether an exception occurs.

Actually, what I am now noticing is that exceptions caught in the outside `try` are the ones that are getting completely ignored, and that probably this should also have something in https://github.com/ansible/ansible/pull/75313/files#diff-a2ea337c649e7371e4acc13677b0c9817a8cda22c1eae619d83c8ff6f330df57R302-R307 to add `process.exception` to `messages` if it exists.

Reply:
I guess you're right... I'm not super familiar with this code, just found it confusing. I'll let somebody else judge the correctness.

OTOH, just from the structuring perspective, if we have a forked subprocess, I'd attempt to make all the unrecoverable exceptions bubble up to the outer shell layer of that subprocess for the final processing and closing of the resources.

Reply:
first, I would log at this point, second make this change of output a flag passed from the caller/initiator to not break existing plugins that expect the current behavior